test/AshToken.js

```diff
@@ -0,0 +1,174 @@
+      const result = await testContract.name();
+      expect(result).to.eql("Ash Token");
+    });
+  });
+
+  describe("totalSupply()", () => {
+    it("has initial supply of 0", async () => {
+      const result = await testContract.totalSupply();
+      expect(result).to.be.a.bignumber.that.equals("0");
+    });
+  });
+
+  describe("balanceOf()", () => {
+    it("has initial balance of 0", async () => {
+      const address = makeRandomAddress();
+      const balance = await testContract.balanceOf(address);
+      expect(balance).to.be.a.bignumber.that.equals("0");
+    });
+  });
+
+  describe("mint()", () => {
+    it("works", async () => {
+      const address = makeRandomAddress();
+      await testContract.mint(address, 7);
+
+      const balance = await testContract.balanceOf(address);
+      expect(balance).to.be.a.bignumber.that.equals("7");
+
+      const totalSupply = await testContract.totalSupply();
+      expect(totalSupply).to.be.a.bignumber.that.equals("7");
+    });
+  });
+
+  describe("transfer()", () => {
+    const owner = accounts[0];
+
+    before(async () => {
+      await testContract.mint(accounts[0], 4);
+    });
+
+    it("works", async () => {
+      const recipient = makeRandomAddress();
+      const { tx } = await testContract.transfer(recipient, 4);
+
+      await expectEvent.inTransaction(tx, ashToken, "Transfer", {
+        // Event argument names not standardized across ERC20 implementation. Use names from contract here.
+        from: owner,
+        to: recipient,
+        value: "4",
+      });
+
+      const balance = await testContract.balanceOf(recipient);
+      expect(balance).to.be.a.bignumber.that.equals("4");
+    });
+
+    it("fails for insufficient balance", async () => {
+      const recipient = makeRandomAddress();
+      await expect(testContract.transfer(recipient, 3000000)).to.be.rejectedWith(/revert/i);
+    });
+  });
+
+  describe("allowance()", () => {
+    it("has default allowance of 0", async () => {
+      const owner = makeRandomAddress();
+      const spender = makeRandomAddress();
+      const allowance = await testContract.allowance(owner, spender);
+      expect(allowance).to.be.a.bignumber.that.equals("0");
+    });
+  });
+
+  describe("approve()", () => {
+    const owner = accounts[0];
+
+    before(async () => {
+      await testContract.mint(owner, 3);
+    });
+
+    it("works", async () => {
+      const spender = makeRandomAddress();
+      const { tx } = await testContract.approve(spender, 3);
+
+      await expectEvent.inTransaction(tx, ashToken, "Approval", {
+        // Event argument names not standardized across ERC20 implementation. Use names from contract here.
+        owner: owner,
+        spender: spender,
+        value: "3",
+      });
+
+      const allowance = await testContract.allowance(owner, spender);
+      expect(allowance).to.be.a.bignumber.that.equals("3");
+    });
+
+    it("works for approval greater than balance", async () => {
```

Comment:
interesting point here

Reply:
I was surprised this works. Initially I wanted to make this an error case.